pkg/rbd/controllerserver.go

```diff
@@ -433,23 +436,66 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		}, nil
 	}
 
+	// this will be used to unreserve snapshot if we failed to get snapshot
+	// information from the ceph backend
+	updatedRbdImage := ""
+	// update the rbd image name to clone from parent volume
+	rbdSnap.RbdImageName = rbdVol.RbdImageName
 	err = reserveSnap(rbdSnap, cr)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer func() {
 		if err != nil {
+			if updatedRbdImage != "" {
+				rbdSnap.RbdImageName = updatedRbdImage
+			}
 			errDefer := undoSnapReservation(rbdSnap, cr)
 			if errDefer != nil {
 				klog.Warningf("failed undoing reservation of snapshot: %s %v", req.GetName(), errDefer)
 			}
 		}
 	}()
 
-	err = cs.doSnapshot(rbdSnap, cr)
+	// generate temp snap struct and vol struct to create a temprory snapshot
+	// and to create a new volume which will be used to take new snapshot
+	tmpSnap := &rbdSnapshot{}
```

Comment:
Perhaps just my opinion, but it's confusing that `rbdSnapshot` and `rbdVolume` are being re-used to describe internal RBD images / snapshots that don't have associated CSI snapshots and volumes. It might be nice to re-org these structures to a parent struct that is only the RBD data and a derived class that adds CSI request related fields.